### ISSUE TITLE
eas update appears to be changing an environment variable in my build

### ISSUE THREAD

**Issue by HM-23-HM** [expo/eas-cli#2236]
### Build/Submit details page URL

https://expo.dev/accounts/hodeem/projects/sheltah/builds/7eb6fb53-0f9a-4ba0-b00d-cba0834d947d 

### Summary

I have `SERVER` set in both the local .env file and the eas.json file. The .env file is in the .gitignore file.
In the JS code, the `SERVER` value is referenced by using `const SERVER = process.env.SERVER`.

If I run `eas build` with a profile, platform etc., the JS code uses the correct `SERVER` value from selected build profile. 

However, when I run `eas update`, the JS code **intermittently** uses the SERVER value that I **only** have in the local .env file.

I say "intermittently" because it appears that the change happens on the first `eas update`, but I can't change the `SERVER` value via. the .env file with any subsequent `eas update`. It has happened twice so far. Can someone let me know what I'm doing wrong or if this is a legitimate bug.

### Managed or bare?

Managed

### Environment

Output from `npx expo-env-info`:
```

  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 11.7.4
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 18.18.1 - ~/.nvm/versions/node/v18.18.1/bin/node
      Yarn: 1.22.19 - ~/.nvm/versions/node/v18.18.1/bin/yarn
      npm: 9.8.1 - ~/.nvm/versions/node/v18.18.1/bin/npm
      Watchman: 2023.08.28.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.2, iOS 15.2, macOS 12.1, tvOS 15.2, watchOS 8.3
    IDEs:
      Android Studio: 2022.1 AI-221.6008.13.2211.9514443
      Xcode: 13.2.1/13C100 - /usr/bin/xcodebuild
    npmPackages:
      @expo/metro-config: ~0.10.0 => 0.10.7 
      babel-preset-expo: ^9.5.2 => 9.5.2 
      expo: ^48 => 48.0.20 
      react: 18.2.0 => 18.2.0 
      react-native: 0.71.14 => 0.71.14 
      react-native-web: ~0.18.10 => 0.18.12 
    npmGlobalPackages:
      eas-cli: 7.2.0
    Expo Workflow: managed
    ```
    
    Output from `npx expo-doctor`:
```
✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check for issues with metro config
✔ Check npm/ yarn versions
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that packages match versions required by installed Expo SDK
✔ Check for legacy global CLI installed locally
✔ Check that native modules do not use incompatible support packages
✔ Check that native modules use compatible support package versions for installed Expo SDK

Didn't find any issues with the project!
```

### Error output

N/A

### Reproducible demo or steps to reproduce from a blank project

1. Run `eas build --auto-submit --profile production --platform android`
2. Run `eas submit --channel production`

Files:
.env
```
SERVER=https://local-domain.com
```

.gitignore
```
node_modules/
.expo/
npm-debug.*
*.jks
*.p8
*.p12
*.key
*.mobileprovision
*.orig.*
web-build/

# macOS
.DS_Store

.env*


android/
ios/
# @generated expo-cli sync-e7dcf75f4e856f7b6f3239b3f3a7dd614ee755a8
# The following patterns were generated by expo-cli

# OSX
#
.DS_Store

# Xcode
#
build/
*.pbxuser
!default.pbxuser
*.mode1v3
!default.mode1v3
*.mode2v3
!default.mode2v3
*.perspectivev3
!default.perspectivev3
xcuserdata
*.xccheckout
*.moved-aside
DerivedData
*.hmap
*.ipa
*.xcuserstate
project.xcworkspace

# Android/IntelliJ
#
build/
.idea
.gradle
local.properties
*.iml
*.hprof

# node.js
#
node_modules/
npm-debug.log
yarn-error.log

# BUCK
buck-out/
\.buckd/
*.keystore
!debug.keystore

# Bundle artifacts
*.jsbundle

# CocoaPods
/ios/Pods/

# Expo
.expo/
web-build/
dist/

# @end expo-cli

dont_upload/

*/Spring Boot
**/build
*.keystore
*.apk

artifacts
```

eas.json
```
{
  "cli": {
    "version": ">= 3.8.1",
    "appVersionSource": "remote"
  },
  "build": {
    "internal": {
      "developmentClient": true,
      "distribution": "internal",
      "ios": {
        "simulator": true,
        "resourceClass": "m-medium",
        "cocoapods": "1.15.2"
      },
      "android": {
        "developmentClient": true
      },
      "channel": "development"
    },
    "staging": {
      "distribution": "store",
      "ios": {
        "resourceClass": "m-medium",
        "cocoapods": "1.15.2",
        "autoIncrement": "buildNumber"
      },
      "android": {
        "resourceClass": "medium",
        "buildType": "app-bundle",
        "autoIncrement": "versionCode"
      },
      "channel": "staging",
      "env": {
        "SERVER": "https://staging-domain.com
      }
    },
    "production": {
      "extends": "staging",
      "channel": "production",
      "env":{
        "SERVER": "https://production-domain.com"
      }
    }
  },
  "submit": {
    "production": {
      "android": {
        "serviceAccountKeyPath": "./dont_upload/google-service-account-key.json",
        "track": "production"
      },
      "ios": {
        "appleId": "***",
        "ascAppId": "***",
        "appleTeamId": "***"
      }
    },
    "staging": {
      "extends": "production",
      "android": {
        "track": "alpha"
      }
    }
  }
}
```

app.config.js
```

module.exports = ({ config }) => ({
  "expo": {
    "name": "***",
    "scheme": "***",
    "slug": "***",
    "version": "1.0.1",
    "orientation": "portrait",
    "icon": "./assets/house_icon.png",
    "userInterfaceStyle": "automatic",
    "splash": {
      "image": "./assets/house_icon.png",
      "resizeMode": "contain",
      "backgroundColor": "#ffffff"
    },
    "assetBundlePatterns": [
      "**/*"
    ],
    "ios": {
      "supportsTablet": true,
      "bundleIdentifier": "***",
      "googleServicesFile": "./config/firebase/GoogleService-Info.plist",
    },
    "android": {
      "package": "***",
      "googleServicesFile": "./config/firebase/google-services.json",
      "config": {
        "googleMaps": {
          "apiKey": '***'
        }
      }
    },
    "web": {
      "favicon": "./assets/favicon.png"
    },
    "extra": {
      "eas": {
        "projectId": "***"
      },
    },
    "plugins": [
      [
        "expo-build-properties",
        {
          "ios": {
            "useFrameworks": "static"
          }
        }
      ],
      "@react-native-firebase/app",
      "@react-native-firebase/crashlytics",
      "@react-native-firebase/perf",
      [
        "expo-image-picker",
        {
          "cameraPermission": "You may need to take photos of your listing with the device camera.",
          "photosPermission": "You may need to upload photos of your listing from the photo gallery.",
        }
      ],
      ["expo-tracking-transparency", {
        "userTrackingPermission": "This identifier will be used to deliver personalized ads to you."
      }],
      "./plugins/android_manifest",
      "./plugins/main_activity"
    ],
    "updates": {
      "url": "***"
    },
    "runtimeVersion": {
      "policy": "sdkVersion"
    },
  },
  ...config
})

```

app.json:
```
{
    "react-native-google-mobile-ads": {
        "android_app_id": "***",
        "ios_app_id": "***",
        "user_tracking_usage_description": "This identifier will be used to deliver personalized ads to you.",
        "sk_ad_network_items": [***],
    }
}
```

**Comment by leons1767** [comment]
`eas update` does not have access to SECRETS variables or environment variables that you defined in eas.json. You can use `dotenv` to export the `.env` variables locally to `eas update`. Obviously, you need to maintain the correct SERVER value in your `.env`.

**Comment by HM-23-HM** [comment]
@leons1767 let's say that I am maintaining the correct SERVER value in the .env file, can you explain why `eas update` changes the value intermittently?

**Comment by leons1767** [comment]
> @leons1767 let's say that I am maintaining the correct SERVER value in the .env file, can you explain why `eas update` changes the value intermittently?

I don't think it changes the value intermittently, it just picks up whatever the SERVER value is when you run `eas update`. You can ensure the right environment variables by loading it from .env using dotenv prior to `eas update`.

**Comment by mkhoussid** [comment]
> > @leons1767 let's say that I am maintaining the correct SERVER value in the .env file, can you explain why `eas update` changes the value intermittently?
> 
> I don't think it changes the value intermittently, it just picks up whatever the SERVER value is when you run `eas update`. You can ensure the right environment variables by loading it from `.env` using `dotenv` prior to `eas update`.

Hello!

Can you please provide a link to the documentation or some sort of basic example? I can't seem to get this to work.

Here's how I'm trying to get this working:

```
cross-env NODE_ENV=production EXPO_PUBLIC_API_URL=https://example.com/api eas update --channel preview --message 'test update'
```
The value for `EXPO_PUBLIC_API_URL` is not being correctly set. It's using the value that I have in my `.env` file (`http://localhost:5000/api`).

**Comment by mkhoussid** [comment]
> > @leons1767 let's say that I am maintaining the correct SERVER value in the .env file, can you explain why `eas update` changes the value intermittently?
> 
> I don't think it changes the value intermittently, it just picks up whatever the SERVER value is when you run `eas update`. You can ensure the right environment variables by loading it from `.env` using `dotenv` prior to `eas update`.

Hello, @leons1767 !

Can you please provide a link to the documentation or some sort of basic example? I can't seem to get this to work.

Here's how I'm trying to get this working:

```
cross-env NODE_ENV=production EXPO_PUBLIC_API_URL=https://example.com/api eas update --channel preview --message 'test update'
```
The value for `EXPO_PUBLIC_API_URL` is not being correctly set. It's using the value that I have in my `.env` file (`http://localhost:5000/api`).

**Comment by mkhoussid** [comment]
I deleted a previous comment, because I found the [relevant documentation](https://docs.expo.dev/eas-update/environment-variables/#sharing-environment-variables-between-local-development-eas-update-and-eas-build) for this issue

**Comment by leons1767** [comment]
> I deleted a previous comment, because I found the [relevant documentation](https://docs.expo.dev/eas-update/environment-variables/#sharing-environment-variables-between-local-development-eas-update-and-eas-build:~:text=Your%20local%20development,eas.json.) for this issue

Glad that you are able to find the answer from the documentation. Have a nice day!

**Comment by Damonio** [comment]
I was breaking my head because I tough that the .env file picked but the problem was the cache, deleting the cache helped me too

```json
    "publish:prod": "cross-env NODE_ENV=production eas update --clear-cache --channel prod --message \"Automated publish to prod\""
```

**Comment by myeljoud** [comment]
> happen to me also, it was intermittenly, because it somehow cached the .env value. so the update needs to be run with `--clear-cache` flag

Perfect! Thanks, I had the same problem and it is now resolved!

**Comment by yourjhay** [comment]
> happen to me also, it was intermittenly, because it somehow cached the .env value. so the update needs to be run with `--clear-cache` flag

This is exactly why.. when changing .env after an update is publish .. it will use the previous value set in .env when doing another publish..

`--clear-cache` works !

**Comment by mayel15** [comment]
> ```json
> NODE_ENV=production eas update --clear-cache
> ```

I faced to the same problem and the `--clear-cache` works for loading env var for `eas update`

**Comment by moritzlang** [comment]
Same happened to me. Ran `eas update` locally and it used a cached `.env` file, which broke production.
Very scary to use this command, therefore set up a github workflow to trigger it in an encapsulated environment.

**Comment by rafael-repasse** [comment]
The thing is, eas update does somehow cache .env variables. --clear-cache fixed it, but this shouldn't happen at all since it doesn't happen with eas run:android for example.

**Comment by OgDev-01** [comment]
I'm actually experiencing this issue... I've tried the `--clear-cache` but to no avail. I have 3 environments variable files for different environments. `.env.development` `.env.staging` and `.env.production`. When i try to send an update to the production channel, i expect it to load values from `.env.production` but it loads from `.env.staging`. I've been on this for two days now, and it's making me delete and create updates here and there just to find a fix. any help will be appreciated 🙏

**Comment by sohammondal** [comment]
> > I'm actually experiencing this issue... I've tried the `--clear-cache` but to no avail. I have 3 environments variable files for different environments. `.env.development` `.env.staging` and `.env.production`. When i try to send an update to the production channel, i expect it to load values from `.env.production` but it loads from `.env.staging`. I've been on this for two days now, and it's making me delete and create updates here and there just to find a fix. any help will be appreciated 🙏
> 
> Finally figured out i was using `APP_ENV` not `NODE_ENV`. changing this in the script fixed the issue for me

Hey 👋 @OgDev-01 can you tell what was the fix? I've similar `.env.XXX` vars in my local and I need to publish an OTA Update.

**Comment by edudepetris** [comment]
alternative: it's using env variables from EAS. As this link describes, running the comman `eas update --environment preview` will read only from EAS and not from your local .env

**Comment by maik-emanoel** [comment]
> happen to me also, it was intermittenly, because it somehow cached the .env value. so the update needs to be run with `--clear-cache` flag

Thanks, man. This solved my problem.

**Comment by daveteu** [comment]
> alternative: it's using env variables from EAS. As this [link](https://docs.expo.dev/eas/environment-variables/#use-environment-variables-with-eas-update) describes, running the comman `eas update --environment preview` will read only from EAS and not from your local .env

I couldn't understand from the docs about Build > EAS Update and got stuck with .env  too. Apparently this `was update --environment preview` works

**Comment by ghost** [comment]
> alternative: it's using env variables from EAS. As this [link](https://docs.expo.dev/eas/environment-variables/#use-environment-variables-with-eas-update) describes, running the comman `eas update --environment preview` will read only from EAS and not from your local .env

thank you

**Comment by MR-AMDEV** [comment]
There is some bug in eas-cli that does not load variables from local env automatically.

**Solution?**
For anyone stuck like me with environment variables not loading from local file and wanna just have exact same behavior to eas build (that loads environment variables from both local env + eas server, just like eas-cli does for **eas build**), do the following:

1. `npm install -g dotenv-cli`
2. `dotenv -e .env -- eas update --environment preview --clear-cache`

Now eas update will load local `.env` variables + variables from preview environment defined on eas server.

**Comment by blxxt** [comment]
Wow, I spent 5+ hours on this exact bug. Found this issue and `--clear-cache` fixed it.

Here's what I run that works: `dotenv -e .env.production -- eas update --clear-cache --channel production`.

The fact that `eas update` caches the envars is pretty wild. Does it really need to be cached? It causes invisible problems by default as people switch btwn updating their production and staging/preview apps. At the very least it should ask or confirm what envars it uses.

**Comment by heymartinadams** [comment]
This happened to us, too, and actually caused a mobile app outage while technicians were working in the field and depended on the mobile app to work.

This is really, really bad design. Whoever implemented putting env variables into the cache and then the cache not clearing on new builds in different environments obviously did not think this through very much.

**Comment by hossam-hani** [comment]
I had this issue too. None of the suggested fixes worked , it only started working after I changed anthing in axios.ts.

**Comment by staz6** [comment]
yeah, this is dumb. You can't rely on this for actual prod live app.

**Comment by noelsoong-simplicity** [comment]
works for me after using --clear-cache, how is this still an issue after 2 years?

**Comment by matheusalisauska** [comment]
`eas update --channel production --environment production` works for me
